exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,247 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusremotewriteexporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/tidwall/wal"
+
+	"go.opentelemetry.io/collector/consumer/consumererror"
+)
+
+type walConfig struct {
+	Dir             string        `mapstructure:"path"`
+	CacheSize       int           `mapstructure:"cache-size"`
+	RefreshDuration time.Duration `mapstructure:"truncate-frequency"`
+}
+
+func (wc *walConfig) createWAL() (*wal.Log, error) {
+	if wc == nil {
+		// There are cases for which the WAL can be disabled.
+		// TODO: Perhaps log that the WAL wasn't enabled.
+		return nil, nil
+	}
+	wal, err := wal.Open(filepath.Join(wc.Dir, "prom_rw"), &wal.Options{
+		SegmentCacheSize: wc.segmentCacheSize(),
+		NoCopy:           true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("prometheusremotewriteexporter: failed to open WAL: %w", err)
+	}
+	return wal, nil
+}
+
+func (wc *walConfig) segmentCacheSize() int {
+	if wc == nil {
+		return 0
+	}
+	if wc.CacheSize <= 0 {
+		return 300
+	}
+	return wc.CacheSize
+}
+
+func (wc *walConfig) refreshDuration() time.Duration {
+	if wc == nil {
+		return 0
+	}
+	if wc.RefreshDuration <= 0 {
+		// Following what Prometheus's WAL defaults.
+		return 1 * time.Minute
+	}
+	return wc.RefreshDuration
+}
+
+func (prwe *PrwExporter) walEnabled() bool { return prwe.walConfig != nil }
+
+var errWALDisabled = errors.New("Write-Ahead-Log is disabled")
+
+func (prwe *PrwExporter) retrieveWALIndices(context.Context) (err error) {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	wal, err := prwe.walConfig.createWAL()
+	if err != nil {
+		return err
+	}
+	prwe.wal = wal
+
+	prwe.rWALIndex, err = prwe.wal.FirstIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
+	}
+	prwe.wWALIndex, err = prwe.wal.LastIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the first WAL index: %w", err)
+	}
+	return nil
+}
+
+// exportThenTruncateWAL uploads the serialized data to the remote write endpoint,
+// and then truncates and retrieves the new indices for the WAL.
+func (prwe *PrwExporter) exportThenTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	// Otherwise, time to flush.
+	if errs := prwe.actualExport(ctx, reqL); len(errs) != 0 {
+		return consumererror.Combine(errs)
+	}
+	if err := prwe.wal.Sync(); err != nil {
```

Comment:
Explain in a comment why we need to Sync.

Reply:
Sure, done, thanks!